### ISSUE TITLE
[java] Improve DoNotCallSystemExit: permit call in main(), flag System.halt

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2157]
**Affects PMD Version:** 
6.20.0.

**Rule:**
[DoNotCallSystemExit](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#donotcallsystemexit)

**Description:**
This rule must make an exception if the exit methods are called from main(). This rule could add a property to dissuade System.exit() being called from any method other than main().

```java

package pmdtests;

import java.util.concurrent.ThreadLocalRandom;

public final class ThrowSwitch {

  private ThrowSwitch() {
    throw new IllegalStateException("Private constructor");
  }

  @SuppressWarnings({"fallthrough", "PMD.MissingBreakInSwitch"})
  public static void main(String... args) {
    int errCode = getSimulatedErrorCode();

    switch (errCode) {
      case 1:
      case 2:
      case 3:
      case 4:
      case 5:
        throw new AssertionError("Error code: " + errCode);
      default:
        System.exit(1);
    }
  }

  public static int getSimulatedErrorCode() {
    return ThreadLocalRandom.current().nextInt(0, 6);
  }

private static void anotherExit() {
System.exit(1);
}
}
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by Fernal73** [comment]
Similarly, the rule should also prevent `Runtime.getRuntime.halt` from being invoked.
Ideally, calling `halt` should flag the violation with a higher severity than normal. Is that possible with PMD?

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#halt(int)

**Comment by adangel** [comment]
> Ideally, calling `halt` should flag the violation with a higher severity than normal. Is that possible with PMD?

Nope, that's not possible - unless we create two separate rules. Then you can overwrite the priority for each separately.

**Comment by Fernal73** [comment]
Any plans to support multiple instances of a rule, differentiated by id, like Checkstyle does? The different configurations then make them two separate rules---in effect.

**Comment by adangel** [comment]
@Fernal73 apart from not being able to configure the same rule twice - there is a fundamental problem: If the rule both covers System.exit and System.halt - we have only one rule that does both. Regardless how often you configure it, it always uses the same configuration for *both* constructs.

If there is really the need to handle System.ext and System.halt differently, then we either need two separate rules or we need the rule to be configurable to either detect only one or both (like adding properties ignoreSystemHalt or ignoreSystemExit).

However, I'm not convinced yet, that we need this additional logic (trying to keep it simple). What's the difference anyway? Either you want to allow terminating the VM in your program or you don't. I don't see, that either of these two methods are less harmful for a web application, so I don't see a point for handling it differently.

I'm more concerned about the name of the rule: Right now, it is called "DoNotCallSystemExit" and it does exactly what it says. But with #2803 we'll extend the very same rule to additionally detect System.halt, so consequently, we should maybe rename the rule as well to "DoNotCallSystemExitOrHalt"? Or more general "DoNotTerminateVM"?

Note: The rule description starts with "Web applications...". If you develop a web application and you are implementing a more complex main method than just calling a web application framework (like [spring boot](https://spring.io/guides/gs/spring-boot/)), then you might have actually a different problem - like you are *not* developing a web application but still have this PMD rule enabled. If you are developing a normal application, there is of course nothing wrong with calling System.exit when you need it. This is just because, if you are developing a web application, your webapp might be run together with other webapps, and if you exit the VM you shutdown all the other webapps that are running in the same application server.

> Any plans to support multiple instances of a rule, differentiated by id, like Checkstyle does?

I keep it in mind, but haven't thought it through yet (like many other things)

**Comment by Fernal73** [comment]
The semantic consequences of System.exit and System.halt differ. Hence, the suggestion. System.halt is more egregious.

DoNotTerminateVM is concise. I like my rule names shorter.

**Comment by Fernal73** [comment]
Is it worth considering a flag that denotes that only command line, web or enterprise rules are to be executed? There would need to be a classification of rules as such. The default category would be all. This would be preferable to creating separate sets of rules for different application types.

**Comment by adangel** [comment]
> Is it worth considering a flag that denotes that only command line, web or enterprise rules are to be executed? There would need to be a classification of rules as such. The default category would be all. This would be preferable to creating separate sets of rules for different application types.

We have some ideas for tagging rules etc. But nothing done yet. See also https://github.com/pmd/pmd/wiki/Schema-version-7.0.0
Tagging would be a way to group rules together. I'd not call it a category, because we have well defined 7 categories already.

I'll rename the rule to *DoNotTerminateVM*, as described here: https://pmd.github.io/latest/pmd_devdocs_rule_deprecation_policy.html#renaming-rules

**Comment by Fernal73** [comment]
> > Is it worth considering a flag that denotes that only command line, web or enterprise rules are to be executed? There would need to be a classification of rules as such. The default category would be all. This would be preferable to creating separate sets of rules for different application types.
> 
> We have some ideas for tagging rules etc. But nothing done yet. See also https://github.com/pmd/pmd/wiki/Schema-version-7.0.0
> Tagging would be a way to group rules together. I'd not call it a category, because we have well defined ~7~ 8 categories already.
> 
> I'll rename the rule to _DoNotTerminateVM_, as described here: https://pmd.github.io/latest/pmd_devdocs_rule_deprecation_policy.html#renaming-rules

I think it's a good idea if users wish to retain all the PMD rules in one humongous xml file but wish to retain the flexibility of executing only the rules they need. A property set within the xml file would not quite serve that purpose. Whether you 'tag' it or 'classify' it is immaterial. Those are implementation details.
